### PR TITLE
dynamic host detection for acme directory

### PR DESCRIPTION
**Pull request by dopey**
Building from #243

### CODE REVIEW

**Comment by maraino** [comment]
Tests are failing.

**Comment by dopey** [comment]
Yeah, I was thinking about passing the context around as well - then I can replace the provisioner.Interface argument in all those methods as well. I'll do that.

I didn't write or fix any unit tests because I wanted to get your feedback first.

File: acme/authority.go [hunk 1]
```diff
@@ -81,16 +103,28 @@ func (a *Authority) GetLink(typ Link, provID string, abs bool, inputs ...string)
...
+func (a *Authority) GetLinkFromBaseURL(typ Link, provID string, abs bool, baseURLFromRequest string, inputs ...string) string {
+	return a.dir.getLinkFromBaseURL(typ, provID, abs, baseURLFromRequest, inputs...)
+}
+
 // GetDirectory returns the ACME directory object.
-func (a *Authority) GetDirectory(p provisioner.Interface) *Directory {
+func (a *Authority) GetDirectory(ctx context.Context) (*Directory, error) {
+	p, err := provisionerFromContext(ctx)
+	if err != nil {
+		fmt.Println("I AM HERE")
```
Comment (clive-jevons) [inline acme/authority.go:116]: This presumably wants to be removed? ;)

File: acme/api/middleware_test.go [hunk 2]
```diff
@@ -28,6 +28,69 @@ func testNext(w http.ResponseWriter, r *http.Request) {
...
 
+func Test_baseURLFromRequest(t *testing.T) {
+	tests := []struct {
+		testFailedDescription string
+		targetURL             string
+		expectedResult        string
+		requestPreparer       func(*http.Request)
+	}{
+		{
+			"HTTP host pass-through failed.",
```
Comment (clive-jevons) [inline acme/api/middleware_test.go:39]: Presumably obsolete because now only https is used?

File: acme/api/middleware_test.go [hunk 3]
```diff
@@ -28,6 +28,69 @@ func testNext(w http.ResponseWriter, r *http.Request) {
...
+		{
+			"HTTPS host pass-through failed.",
+			"https://my.dummy.host",
+			"https://my.dummy.host",
+			nil,
+		},
+		{
+			"Port pass-through failed",
+			"http://host.with.port:8080",
+			"http://host.with.port:8080",
```
Comment (clive-jevons) [inline acme/api/middleware_test.go:53]: Probably needs changing to https to succeed?

File: acme/api/middleware_test.go [hunk 4]
```diff
@@ -28,6 +28,69 @@ func testNext(w http.ResponseWriter, r *http.Request) {
...
+			nil,
+		},
+		{
+			"Explicit host from Request.Host was not used.",
+			"http://some.target.host:8080",
+			"http://proxied.host",
```
Comment (clive-jevons) [inline acme/api/middleware_test.go:59]: Probably needs changing to https to succeed?

File: acme/api/middleware_test.go [hunk 5]
```diff
@@ -28,6 +28,69 @@ func testNext(w http.ResponseWriter, r *http.Request) {
...
+			func(r *http.Request) {
+				r.Host = "proxied.host"
+			},
+		},
+		{
+			"Explicit forwarded protocol from request header X-Forwarded-Proto was not used.",
```
Comment (clive-jevons) [inline acme/api/middleware_test.go:65]: Does this test still make sense? I think we're forcing https now?

File: acme/authority_test.go [hunk 6]
```diff
@@ -82,6 +82,20 @@ func TestAuthorityGetDirectory(t *testing.T) {
 	assert.Equals(t, acmeDir.KeyChange, fmt.Sprintf("https://ca.smallstep.com/acme/%s/key-change", URLSafeProvisionerName(prov)))
 }
 
+func TestAuthorityGetDirectoryWithBaseURL(t *testing.T) {
+	baseURL := "http://my.proxied.host"
```
Comment (clive-jevons) [inline acme/authority_test.go:86]: Does this need changing to https?

File: acme/api/handler_test.go [hunk 7]
```diff
@@ -276,6 +276,7 @@ func TestHandlerGetDirectory(t *testing.T) {
 		t.Run(name, func(t *testing.T) {
 			h := New(auth).(*Handler)
 			req := httptest.NewRequest("GET", url, nil)
+			req.Header.Add("X-Forwarded-Proto", "https")
```
Comment (clive-jevons) [inline acme/api/handler_test.go:279]: Is this still required? I think we're now ignoring that header value?

**Comment by dopey** [comment]
@maraino tests fixed

File: acme/directory.go [hunk 8]
```diff
@@ -114,7 +120,11 @@ func (d *directory) getLink(typ Link, provisionerName string, abs bool, inputs .
 		link = fmt.Sprintf("/%s/%s/%s/finalize", provisionerName, OrderLink.String(), inputs[0])
 	}
 	if abs {
-		return fmt.Sprintf("https://%s/%s%s", d.dns, d.prefix, link)
+		baseURL := baseURLFromRequest
+		if baseURL == "" {
+			baseURL = "https://" + d.dns
+		}
+		return fmt.Sprintf("%s/%s%s", baseURL, d.prefix, link)
```
Comment (maraino) [inline acme/directory.go:127]: Is there any need to url-encode things in this method. If there is I'll recommend you to use url.URL.

**Comment by ThreeFx** [comment]
FWIW I just ran into this issue and would be happy to test :)

**Comment by dopey** [comment]
@ThreeFx should be merging this to master shortly (today). Just waiting on a final review. Should make it into a release within a few weeks.

File: acme/directory.go [hunk 9]
```diff
@@ -119,12 +125,26 @@ func (d *directory) getLinkFromBaseURL(typ Link, provisionerName string, abs boo
...
+
 	if abs {
-		baseURL := baseURLFromRequest
-		if baseURL == "" {
-			baseURL = "https://" + d.dns
+		// Copy the baseURL value from the pointer. https://github.com/golang/go/issues/38351
+		u := url.URL{}
+		if baseURL != nil {
+			u = *baseURL
+		}
```
Comment (maraino) [inline acme/directory.go:134]: `url.URL` is usually used as a pointer, so it might be more comfortable to do `u := new(url.URL)`

Comment (dopey) [reply acme/directory.go:134]: Doing this so as not to change the value inside baseURL.

**Pull request merged**